anchore_engine/clients/grype_wrapper.py

```diff
@@ -455,3 +456,61 @@ def query_vulnerabilities(
             return query.all()
         finally:
             read_lock.release()
+
+
+@dataclass
+class RecordSource:
+    count: int
+    feed: str
+    group: str
+    last_synced: str
+
+
+def query_record_source_counts():
+    """
+    Query the current feed group counts for all current vulnerabilities.
+    """
+    # Get and release read locks
+    read_lock = grype_db_lock.gen_rlock()
+    if read_lock.acquire(blocking=False, timeout=60):
+        try:
+            logger.debug("Querying grype_db for feed group counts")
+
+            # Get the counts for each record source
+            results = (
+                _get_grype_db_session()
```

Comment:
db sessions should also be handled using a context manager. we do this currently for db sessions to our persistence context

Reply:
@nightfurys One the first point, I looked into doing that when I first pulled in that library, and due to the way the authors did the inheritance structure, you can't parameterize the the acquire call via a context manager through call to the __enter__(). I'd need to add something extra (Vijay also suggested that) to wrap the calls to the lock in order to use a context manager with it.

On the second point, should I be closing and recreating the session object on every query then? I.e. the SessionMaker is what gets created and persisted when the grypeDB changes? I was thinking since we are using grype db as a read-only db it was ok to leave that open (since the underlying data being queried won't change), but it occurs to me now that there's probably a performance cost to leaving that open.

Reply:
If that is the case I'm happy to make that change, but I think I'd rather do that in [https://github.com/anchore/anchore-engine/pull/1029](https://github.com/anchore/anchore-engine/pull/1029) too. I'm already changing how that variable is handled there, so it will be easier to change what its storing there too (rather than doing it here and then having a messier rebase of that branch on this one).

Reply:
@nightfurys I thought more about what you and @Vijay-P were saying here about context managers for the lock and the db session. I've got implementations of each I am testing locally, but I'm going to target them for #1029 rather than this PR.